### ISSUE TITLE
on_train_batch_end in pl.LightningModule returns mutated value under 'loss' key when using gradient accumulation

### ISSUE THREAD

**Issue by spaghettiSystems** [Lightning-AI/pytorch-lightning#18464]
### Bug description

Title.

In the documentation it says

```
LightningModule.on_train_batch_end(outputs, batch, batch_idx)

 (Union[Tensor, Dict[str, Any]]) – The outputs of training_step(x)
 ```
 
 Which does not mention any mutation caused by gradient accumulation. It is my opinion that dividing the loss by the number of gradient accumulation steps should be invisible to the user, but changing the documentation should be OK too.
![image](https://github.com/Lightning-AI/lightning/assets/58869888/8ed39878-b1df-442d-98ad-f8238059d48a)

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

add `accumulator = GradientAccumulationScheduler(scheduling={0:1, 1:100)` to pl.Trainer callbacks. Start logging the loss in LightningModule's on_train_batch_end function. 

You will see that after 1 epoch AGI will be achieved.

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

more info: I wasted 1 day trying to debug this "very strange overfitting" 

cc @borda
